pkg/http/sender.go

```diff
@@ -0,0 +1,660 @@
+// each ILP message. Instead, the messages should be written in
+// batches followed by a Flush call. Optimal batch size may vary
+// from 100 to 1,000 messages depending on the message size and
+// configured buffer capacity.
+func (s *LineSender) Flush(ctx context.Context) error {
+	var (
+		req           *http.Request
+		retryInterval time.Duration
+
+		maxRetryInterval = time.Second
+	)
+
+	if s.closed {
+		return errors.New("cannot flush a closed LineSender")
+	}
+
+	err := s.LastErr()
+	s.ClearLastErr()
+	if err != nil {
+		s.DiscardPendingMsg()
+		return err
+	}
+	if s.HasTable() {
+		s.DiscardPendingMsg()
+		return errors.New("pending ILP message must be finalized with At or AtNow before calling Flush")
+	}
+
+	// timeout = ( request.len() / min_throughput ) + request_timeout
+	// nb: conversion from int to time.Duration is in milliseconds
+	timeout := time.Duration(s.Len()/s.minThroughputBytesPerSecond)*time.Second + s.requestTimeout
+	reqCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	req, err = http.NewRequestWithContext(
+		reqCtx,
+		http.MethodPost,
+		s.uri,
+		s,
+	)
+	if err != nil {
+		return err
+	}
+
+	if s.user != "" && s.pass != "" {
+		req.SetBasicAuth(s.user, s.pass)
+	}
+
+	if s.token != "" {
+		if req.Header.Get("Authorization") != "" {
+			return errors.New("cannot use both Basic and Token Authorization")
+		}
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", s.token))
+	}
+
+	resp, err := s.client.Do(req)
+	if err == nil {
+		defer resp.Body.Close()
+
+		// If the requests succeeds with a non-error status code, flush has succeeded
+		if !isRetryableError(resp.StatusCode) {
+			s.msgCount = 0
+			return nil
+		}
+		// Otherwise, we will retry sending the request until the retry timeout is breached
+		err = fmt.Errorf("Non-OK Status Code %d: %s", resp.StatusCode, resp.Status)
+	}
+
+	if s.retryTimeout > 0 {
+		retryInterval = 10 * time.Millisecond
+		for err != nil {
+			jitter := time.Duration(rand.Intn(10)) * time.Millisecond
+			time.Sleep(retryInterval + jitter)
+
+			resp, retryErr := s.client.Do(req)
+			if retryErr == nil {
+				defer resp.Body.Close()
+
+				// If the requests succeeds with a non-error status code, flush has succeeded
+				if !isRetryableError(resp.StatusCode) {
+					s.msgCount = 0
+					return nil
+				}
+				retryErr = fmt.Errorf("Non-OK Status Code %d: %s", resp.StatusCode, resp.Status)
+			}
+
+			// If the error is an http timeout (due to the context expiring),
+			// return the previously-encountered error
+			var urlErr *url.Error
+			if errors.As(retryErr, &urlErr) {
+				if urlErr.Timeout() {
+					return err
+				}
+			}
+
+			err = retryErr
+
+			// If the attempt fails, retry with exponentially-increasing timeout
+			// up to a global maximum (1 second)
+			retryInterval = retryInterval * 2
+			if retryInterval > maxRetryInterval {
+				retryInterval = maxRetryInterval
+			}
+		}
+	}
+
```

Comment:
I moved the msgCount implementation to the buffer so we can update it with lastMsgPos.  But I had a thought, what do we do with message count when `buffer.WriteTo()` returns an error and leaves a partially-written buffer? We correctly decrement lastMsgPos, but there's no way to tell how many messages were successfully written.

Reply:
In that case, we could reset sender's state since such an error means that the connection got broken. In case of TCP, the sender should be closed by user. In case of HTTP, the sender may be reused to write messages once again and try sending them.